go-controller/pkg/ovn/kubevirt.go

```diff
@@ -110,18 +121,263 @@ func (oc *DefaultNetworkController) deleteDHCPOptions(pod *kapi.Pod) error {
+	if err != nil {
+		return fmt.Errorf("failed cleaning up VM when checking if pod is leftover: %v", err)
+	}
+	// Everything has already being cleand up since this is and old migration
+	// pod
+	if isMigratedSourcePodStale {
+		return nil
+	}
+	// This pod is not part of ip migration so we don't need to clean up
+	if !kubevirt.IsPodLiveMigratable(pod) {
+		return nil
+	}
+	if err := oc.deleteDHCPOptions(pod); err != nil {
+		return err
+	}
+	if err := oc.deleteStaticRoutesForMigratedPod(pod); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (oc *DefaultNetworkController) ensureLocalZonePodAddressesToNodeRoute(pod *kapi.Pod) error {
+	podAnnotation, err := util.UnmarshalPodAnnotation(pod.Annotations, "default")
+	if err != nil {
+		return fmt.Errorf("failed reading ovn annotation: %v", err)
+	}
+	switchNames, err := oc.getSwitchNames(pod)
+	if err != nil {
+		return fmt.Errorf("failed configuring pod routing when getting switch current and original name: %v", err)
+	}
+
+	// VM is running at the node that owns the subnet the point to point
+	// routing is not needed
+	if switchNames.Current == switchNames.Original {
+		// Delete not needed policies or static routes
+		if err := libovsdbops.DeleteLogicalRouterPoliciesWithPredicate(oc.nbClient, types.OVNClusterRouter, func(policy *nbdb.LogicalRouterPolicy) bool {
+			policyVMLabel, ok := policy.ExternalIDs[string(libovsdbops.VirtualMachineIndex)]
+			return ok && policyVMLabel == pod.Labels[kubevirtv1.VirtualMachineNameLabel]
+		}); err != nil {
+			return fmt.Errorf("failed configuring pod routing when deleting stale policies for pod %s/%s: %v", pod.Namespace, pod.Name, err)
+		}
+		if err := libovsdbops.DeleteLogicalRouterStaticRoutesWithPredicate(oc.nbClient, types.OVNClusterRouter, func(route *nbdb.LogicalRouterStaticRoute) bool {
+			routeVMLabel, ok := route.ExternalIDs[string(libovsdbops.VirtualMachineIndex)]
+			return ok && routeVMLabel == pod.Labels[kubevirtv1.VirtualMachineNameLabel]
+		}); err != nil {
+			return fmt.Errorf("failed configuring pod routing when deleting stale static routes for pod %s/%s: %v", pod.Namespace, pod.Name, err)
+		}
+		return nil
+	}
+	lrpName := types.GWRouterToJoinSwitchPrefix + types.GWRouterPrefix + pod.Spec.NodeName
+	lrpAddresses, err := util.GetLRPAddrs(oc.nbClient, lrpName)
+	if err != nil {
+		return fmt.Errorf("failed configuring pod routing when reading LRP %s addresses: %v", lrpName, err)
+	}
+
+	// For IC this route will match the cluster CIDR and route traffic to
+	// the gw, the VM's live migrated IP will match there when they
+	// are migrated at a node not owning the VM subnet
+	if config.OVNKubernetesFeature.EnableInterconnect {
+		for _, clusterSubnet := range config.Default.ClusterSubnets {
+			nodeGRAddress, err := util.MatchFirstIPNetFamily(utilnet.IPFamilyOfCIDR(clusterSubnet.CIDR) == utilnet.IPv6, lrpAddresses)
+			if err != nil {
+				return err
+			}
+			clusterSubnetRoute := nbdb.LogicalRouterStaticRoute{
```

Comment:
I am not sure this would work with multiple nodes in a single zone. In that case you would need the specific policy like in the non-IC case? Why not do that anyway?

Reply:
On that scenario I think it may not work or it will introduce some inter node communication since egress traffic will be routed over a different node where the pod is running. 

The idea was to reduce the number of "moving" parts, a cluster wide route that don't affect the non live migrated pod is ideal on that regard.

Is it possible to detect if we are on a "pure" interconnect scenario or a hybrid one ? so we can configure this one way or another.

Reply:
Why do you say `egress traffic will be routed over a different node than the one the pod is running`? If you add the same policy you are using for the non-interconnect case, that should not happen?

Reply:
Anyhow after going back and forth with @jcaamano looks like to support multiple nodes per zone and live migration we have to add at least one policy/route per VM meaning adding a new flow at each of the nodes at the zone per running VM.

In the case of hypershift we are going to have one node per zone so the cluster wide static route is good enough for us.

So we have three options:
- Limit live migration to NO IC or one node per zone IC and we support it at a follow up:
  - **Pros**: less flows at local zone node
  - **Cons**: migration to the follow up change
 - Support multiple nodes per ZONE at live migration by adding more policy/routes
   - **Pros**: Do not have to detect if there are multiple nodes on the zone, support dynamically adding nodes to the zone
   - **Cons**: at least one more policy/route per VM at local zone nodes.
 - Detect if IC support multiple node per zone and add the extra policy/route or the cluster wide cidr route
   - **Pros**: We have the best of both worlds
   - **Cons**: The detection mechanism is not desirable.
  
@trozet what do  you think about it ?

Reply:
I think we should go with option 2, where we add a route per IP. There are a few reasons why:

1. You are already adding a per IP route according to the output we referred to before:
```
--->         10.244.2.8               10.244.2.8 dst-ip node-vm-output-port
               100.64.0.2               168.254.0.2 dst-ip
               100.64.0.3               168.254.0.3 dst-ip
               100.64.0.4                100.64.0.4 dst-ip
            10.244.0.0/24               168.254.0.3 dst-ip
            10.244.2.0/24               168.254.0.2 dst-ip
            10.244.1.0/24                100.64.0.4 src-ip
--->      10.244.0.0/16                100.64.0.4 src-ip
```

So I think it is cleaner to look at the routes configured in the code and here and know what belongs to what. It also makes cleanup easier.

2. I'm a little concerned about having this generic 10.244.0.0/16 route configured always. I get that it works today and is a slick way to minimize configuration to achieve what we want...but I'm a little wary that in the future it may accidentally route traffic somewhere we did not expect.

3. The multiple nodes per zone thing is broken by this change. I don't think we have any plans to support multiple nodes per zone downstream, but upstream I guess we support it for now. I'm not sure if this is something we want to continue to support upstream or not. I guess that should be a separate discussion in the community meeting. @numansiddique FYI. Either way I want to avoid adding yet another matrix of this thing isnt supported in one mode, but supported in others.

4. I think adding routes per pod O(n * 2) could become a perf/scale issue down the road. We should watch out for this and if it is something that costs us perf/scale wise then we should consider moving back to the single generic route, but until we know for sure it is problem I dont think we should worry about it.

Reply:
Ok, let's add suport for live-migration and multiple nodes per zone, at least we know it can be simplified in the future.

Reply:
To implement egress traffic like we do without interconnect using a policy we need a static route to match if we don't do that policy are not evaluated, but we still need to match the e/w traffic routes so this "dummy" static route has to be a /24 or /16 route src-ip since it src-ip has less priority than dst-ip, this leave us with this options:

1.- A cluster wide one -> `10.244.0.0/16 0.0.0.0 rtoj-ovn_cluster_router` 
2.- A per subnet one -> `10.244.2.0/24 0.0.0.0 rtoj-ovn_cluster_roter`
3.- A per VM one (ecmp) -> `10.244.2.0/24 10.244.2.8 rtoj-ovn_cluster_router`

If we don't set the output port northd fails since it's not able to discover the port from the nexthop

Reply:
At the end we are going with option 2 but replacing blackhole nexthop with transit switch port address